lib/linter/linter.js

```diff
@@ -778,14 +779,31 @@ function stripUnicodeBOM(text) {
 /**
  * Get the options for a rule (not including severity), if any
  * @param {Array|number} ruleConfig rule configuration
+ * @param {Object|undefined} defaultOptions rule.meta.defaultOptions
  * @returns {Array} of rule options, empty Array if none
  */
-function getRuleOptions(ruleConfig) {
+function getRuleOptions(ruleConfig, defaultOptions) {
     if (Array.isArray(ruleConfig)) {
-        return ruleConfig.slice(1);
+        return deepMergeArrays(defaultOptions, ruleConfig.slice(1));
```

Comment:
In flat config case, `defaultOptions` have already been merged during the validation so this would merge them again?

Reply:
Aha! Yes, you're right. And there's an additional bug, I think, that default values from schemas aren't being applied in the branch right now. 0dc481013 has a passing unit test for flat config and a failing unit test for legacy config.

```plaintext
1) Linter
       options
         rules should apply meta.defaultOptions on top of schema defaults:

      AssertionError: expected { Object (inBoth, inDefaultOptions) } to deeply equal { Object (inBoth, inDefaultOptions, ...) }
      + expected - actual

       {
         "inBoth": "from-default-options"
         "inDefaultOptions": "from-default-options"
      +  "inSchema": "from-schema"
       }
```

I _think_ what needs to happen is moving the validation so it happens once for each config, and with both meta-based and schema-based default options applied. Working on that now.

---

Adding my notes as I found the code flow hard to keep track of without them:
* `Linter` takes in a `configType` of type `"eslintrc" | "flat"`
    * For `"eslintrc"`, it calls `this._verifyWithConfigArray` -> ... -> `_verifyWithoutProcessors` -> `runRules`
    * For `"flat"`, it calls `this._verifyWithFlatConfigArray` -> ... -> `_verifyWithFlatConfigArrayAndWithoutProcessors` -> `runRules`
        * In `_verifyWithFlatConfigArrayAndWithoutProcessors`, `const ruleValidator = new RuleValidator()` calls to a `validate()` method. That uses a `this.validators` to build validators from `ajv.compile(schema)`. This applies **schema** default options.
        * Later in `validate()`, `deepMergeArrays` is called with `deepMergeArrays`. This applies **meta** default options.
* `runRules` calls `getRuleOptions(configuredRules[ruleId], rule.meta?.defaultOptions)`. This applies **meta** default options.

With both flat and legacy configs, rule validation happens inside `runRules`.

With flat configs, rule validation additionally happens _before_ `runRules`. This earlier validation is applied to the result of merging both **meta** and **schema** default options.

With legacy configs, only the `runRules` validation is applied. That's why it doesn't apply to **schema** default options.

Reply:
OK, sorry, I'm thoroughly confused and think I need help here. On `main`, `Linter`'s unit tests never apply **schema** defaults - neither for flat nor legacy config. https://github.com/JoshuaKGoldberg/eslint/commit/9ccaa85d81f6b108cf92bcea458dc80437698b6a shows a passing unit test per config type showing schema defaults not being applied.

In the flat config code path, `Linter` calls to `ruleValidator.validate(...)`. That means Ajv's options validation and **schema** defaulting applies there as well as in `flat-config-array.js`.

So now I'm unsure what the intended boundaries are supposed to be. Do we want **schema** defaulting to be moved to happen inside `Linter` now? Is that a breaking change?

Reply:
Just to clarify what happens with schema defaults.

Schema defaults in the top-level array never apply. That is, I believe, an unintentional omission (i.e., a bug) in the code that handles provided options. We could have considered fixing that for v9 if we hadn't decided to switch to `defaultOptions` and eventually drop schema defaults. So, in https://github.com/JoshuaKGoldberg/eslint/commit/9ccaa85d81f6b108cf92bcea458dc80437698b6a, `default: { inSchema: "from-schema-root" }` is a no-op. On the other hand, `inSchema: { default: "from-schema", type: "string" }` can apply, but you'd need to pass `"my-rule": ["error", {}]`. This should work in config _files_, both eslintrc and flat config. This should also work when `Linter` is used directly in flat config mode. But it won't work when `Linter` is used directly in eslintrc mode, because in eslintrc mode Linter doesn't validate rule options, which is probably another unintentional omission. I'm not sure if it's worth fixing that at this point, because eslintrc mode is deprecated and will be dropped in v10.

Reply:
Got it, thanks - that's very helpful.

> should work in config _files_, both eslintrc and flat config
> won't work when `Linter` is used directly in eslintrc mode

Just confirming then - is the behavior in the latest commit, adb7972d9, an acceptable one for v9? That its `Linter` unit test for flat config applies on top of schema defaults, while its unit test for legacy config ignores schema defaults?

Reply:
I think if we are maintaining the current eslintrc behavior, then we are okay. @mdjermanovic?

Reply:
ping @mdjermanovic

Reply:
I think it's okay to keep the current behavior regarding validation and schema defaults when `Linter` is used directly in eslintrc mode. That is, rule options passed in the config are not validated and JSON schema defaults are not applied to them. We'd probably want to change this if the eslintrc config system wasn't deprecated.

But the new feature `meta.defaultOptions` should apply when `Linter` is used directly in eslintrc mode.